cpukit/libnetworking/libc/addr2ascii.3
.\"
.\" Copyright 1996 Massachusetts Institute of Technology
.\"
.\" Permission to use, copy, modify, and distribute this software and
.\" its documentation for any purpose and without fee is hereby
.\" granted, provided that both the above copyright notice and this
.\" permission notice appear in all copies, that both the above
.\" copyright notice and this permission notice appear in all
.\" supporting documentation, and that the name of M.I.T. not be used
.\" in advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.  M.I.T. makes
.\" no representations about the suitability of this software for any
.\" purpose.  It is provided "as is" without express or implied
.\" warranty.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY M.I.T. ``AS IS''.  M.I.T. DISCLAIMS
.\" ALL EXPRESS OR IMPLIED WARRANTIES WITH REGARD TO THIS SOFTWARE,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT
.\" SHALL M.I.T. BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
.\" USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
.\" OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	$ANA: addr2ascii.3,v 1.1 1996/06/13 18:41:46 wollman Exp $
.\"     $Id$
.\"
.Dd June 13, 1996
.Dt ADDR2ASCII 3
.Os
.Sh NAME
.Nm addr2ascii ,
.Nm ascii2addr
.Nd Generic address formatting routines
.Sh SYNOPSIS
.Fd #include <sys/types.h>
.Fd #include <netinet/in.h>
.Fd #include <arpa/inet.h>
.Ft "char *"
.Fn addr2ascii "int af" "const void *addrp" "int len" "char *buf"
.Ft int
.Fn ascii2addr "int af" "const char *ascii" "void *result"
.Sh DESCRIPTION
The routines
.Fn addr2ascii
and
.Fn ascii2addr
are used to convert network addresses between binary form and a
printable form appropriate to the address family.  Both functions take
an
.Fa af
argument, specifying the address family to be used in the conversion
process.
(Currently, only the
.Dv AF_INET
and
.Dv AF_LINK
address families are supported.)
.Pp
The
.Fn addr2ascii
function
is used to convert binary, network-format addresses into printable
form.  In addition to
.Fa af ,
there are three other arguments.  The
.Fa addrp
argument is a pointer to the network address to be converted.
The
.Fa len
argument is the length of the address.  The
.Fa buf
argument is an optional pointer to a caller-allocated buffer to hold
the result; if a null pointer is passed,
.Fn addr2ascii
uses a statically-allocated buffer.
.Pp
The
.Fn ascii2addr
function performs the inverse operation to
.Fn addr2ascii .
In addition to
.Fa af ,
it takes two parameters,
.Fa ascii
and
.Fa result .
The
.Fa ascii
parameter is a pointer to the string which is to be converted into
binary.  The
.Fa result
parameter is a pointer to an appropriate network address structure for
the specified family.
.Pp
The following gives the appropriate structure to use for binary
addresses in the specified family:
.Pp
.Bl -tag -width AF_INETxxxx -compact
.It Dv AF_INET
.Li struct in_addr
.Pq in Aq Pa netinet/in.h
.It Dv AF_LINK
.Li struct sockaddr_dl
.Pq in Aq Pa net/if_dl.h
.\" .It Dv AF_INET6
.\" .Li struct in6_addr
.\" .Pq in Aq Pa netinet6/in6.h
.El
.Sh RETURN VALUES
The
.Fn addr2ascii
function returns the address of the buffer it was passed, or a static
buffer if the a null pointer was passed; on failure, it returns a null
pointer.
The
.Fn ascii2addr
function returns the length of the binary address in bytes, or -1 on
failure.
.Sh EXAMPLES
The
.Xr inet 3
functions
.Fn inet_ntoa
and
.Fn inet_aton
could be implemented thusly:
.Bd -literal -offset indent
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

char *
inet_ntoa(struct in_addr addr)
{
	return addr2ascii(AF_INET, &addr, sizeof addr, 0);
}

int
inet_aton(const char *ascii, struct in_addr *addr)
{
	return (ascii2addr(AF_INET, ascii, addr) 
	    == sizeof(*addr));
}
.Ed
.Pp
In actuality, this cannot be done because
.Fn addr2ascii
and
.Fn ascii2addr
are implemented in terms of the
.Xr inet 3
functions, rather than the other way around.
.Sh ERRORS
When a failure is returned,
.Li errno
is set to one of the following values:
.Bl -tag -width [EPROTONOSUPPORT]
.It Bq Er ENAMETOOLONG
The
.Fn addr2ascii
routine was passed a
.Fa len
parameter which was inappropriate for the address family given by
.Fa af .
.It Bq Er EPROTONOSUPPORT
Either routine was passed an
.Fa af
parameter other than
.Dv AF_INET
or
.Dv AF_LINK .
.It Bq Er EINVAL
The string passed to
.Fn ascii2addr
was improperly formatted for address family
.Fa af .
.El
.Sh SEE ALSO
.Xr inet 3 ,
.Xr linkaddr 3 ,
.Xr inet 4
.Sh HISTORY
An interface close to this one was originally suggested by Craig
Partridge.  This particular interface originally appeared in the
.Tn INRIA
.Tn IPv6
implementation.
.Sh AUTHORS
Code and documentation by
.An Garrett A. Wollman ,
MIT Laboratory for Computer Science.
.Sh BUGS
The original implementations supported IPv6.  This support should
eventually be resurrected.  The
.Tn NRL
implementation also included support for the
.Dv AF_ISO
and
.Dv AF_NS
address families.
.Pp
The genericity of this interface is somewhat questionable.  A truly
generic interface would provide a means for determining the length of
the buffer to be used so that it could be dynamically allocated, and
would always require a
.Dq Li "struct sockaddr"
to hold the binary address.  Unfortunately, this is incompatible with existing
practice.  This limitation means that a routine for printing network
addresses from arbitrary address families must still have internal
knowledge of the maximum buffer length needed and the appropriate part
of the address to use as the binary address.


ChangeLog
2006-11-18	Ralf Corsépius <ralf.corsepius@rtems.org>

	* bootstrap: Add -p (regenerate preinstall.ams).

2006-11-15	Ralf Corsépius <ralf.corsepius@rtems.org>

	* ampolish3: New.

2006-10-17	Ralf Corsépius <ralf.corsepius@rtems.org>

	* configure.ac: Requires automake-1.10. Require autoconf-2.60.
	* aclocal/rtems-top.m4, aclocal/version.m4: 
	Adaptions to automake-1.10.

2006-09-13	Ralf Corsépius <ralf.corsepius@rtems.org>

	* bspkit: Remove.

2006-08-09	Kolja Waschk <waschk@telos.de>

	* config-ml.in: New port to Altera NIOS II.

2006-07-09      Thomas Doerfler <Thomas.Doerfler@embedded-brains.de>

	* aclocal/bsp-alias.m4: (PR1117) Add brs5l as alias for gen5200
	* aclocal/check-bsps.m4: (PR1117) Add brs5l as alias for gen5200
	
2006-06-23	Worth Burruss <worth@motioncontrol.org>

	* README.configure: Add mcf5235.

2006-04-18	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to rtems-4.6.99.3.

2006-04-18  Joel Sherrill <joel@OARcorp.com>

	* aclocal/version.m4: Updated to rtems-4.6.99.3.

2006-04-18	Joel Sherrill <joel@OARcorp.com>

	* release_support: Fix typos.

2006-04-18	Joel Sherrill <joel@OARcorp.com>

	* ChangeLog, release_support: Fix typos.

2006-03-09	Ralf Corsepius <ralf.corsepius@rtems.org>

	* acinclude.m4: Requires autoconf-2.59.

2006-03-07	Joel Sherrill <joel@OARcorp.com>

	PR 694/rtems
	* release_support: Do not include .cvsignore files in release tarballs.

2006-02-01	Ralf Corsepius <ralf.corsepius@rtems.org>

	PR 882/make_build
	* bootstrap: Only remove bspopts.h.in instead of bspopts.h*.

2006-01-12	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/bsp-alias.m4: Revert Joel's 2006-01-10 change.
	Alpha-sort the aliases list.
	* automake/host.am, automake/local.am, automake/subdirs.am:
	Remove depend-am, depend-recursive.

2006-01-10	Joel Sherrill <joel@OARcorp.com>

	* aclocal/bsp-alias.m4: Add sis as alias of erc32.

2006-01-09	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/bsp-alias.m4: Remove brs5l. Remove leon2.
	Alias leon1 to leon2. Add sis.
	* aclocal/check_bsps.m4: Remove brs5l.
	Remove leon. Add leon2. Add sis to erc32 family.

2005-11-31	Ralf Corsepius <ralf.corsepius@rtems.org>

	PR 851/bsps
	* aclocal/bsp-alias.m4, aclocal/check-bsps.m4: Add gen5200 BSP
	family.

2005-11-03	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/check-bsps.m4: Add the plethora of mbx* BSP.

2005-09-02	Joel Sherrill <joel@OARcorp.com>

	* VERSION: Updated to rtems-4.6.99.2.

2005-09-02  Joel Sherrill <joel@OARcorp.com>

	* aclocal/version.m4: Updated to rtems-4.6.99.2.

2005-07-06      Philippe Simons

	* MAINTAINERS: Added me as arm/gp32 bsp maintainer.

2005-04-28	Joel Sherrill <joel@OARcorp.com>

	* README.configure: Updated to account for removal of dmv177 and
	ppcn_60x as well as other additions and deletions.

2005-02-04	Ralf Corsepius <ralf.corsepius@rtems.org>

	* aclocal/enable-itron.m4: Allow unix.

2005-01-02	Ralf Corsepius <ralf.corsepius@rtems.org>

	* Makefile.am: Remove ACLOCAL_FILES.

2004-11-16	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/check-bsps.m4: Use ls instead of globs to pickup
	bsp_specs.

2004-11-12	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/check-bsps.m4: Add MVME2100 BSP as alias of
	motorola_shared.

2004-11-10	Joel Sherrill <joel@oarcorp.com>

	* aclocal/bsp-alias.m4: Add MVME2100 BSP as alias of motorola_shared.

2004-10-22	Ralf Corsepius <ralf_corsepius@rtems.org>

	* alocal/check-custom-bsp.m4: Reflect new location of bspkit*cfg.

2004-10-18	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/check-bsps.m4: Add bspkit support. Don't check for
	presence of *.cfg.
	* aclocal/check-custom-bsp.m4: Add bspkit support.

2004-10-18	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/enable-bare.m4: Remove.

2004-09-29	Joel Sherrill <joel@OARcorp.com>

	* MAINTAINERS, README.configure, c/ACKNOWLEDGEMENTS,
	c/TOOL_TARGETS: i960 obsoleted and all references removed.

2004-09-24	Ralf Corsepius <ralf_corsepius@rtems.org>

	* configure.ac: Require automake > 1.9.

2004-09-24	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/rtems-cpu-subdirs.m4: Remove (unused).

2004-08-18	Ralf Corsepius <ralf_corsepius@rtems.org>

	PR 673/make_build
	* config-ml.in: Quote 'cd "${ml_origdir}"'.
	Quote 'cd "${ML_POPDIR}"'.

2004-08-17	Ralf Corsepius <ralf_corsepius@rtems.org>

	PR 673/make_build
	* acinclude.m4 (_RTEMS_CONFIG_SUBDIR): Quote ac_popdir uses.

2004-07-14      Joel Sherrill <joel@OARcorp.com>

        * MAINTAINERS: Added Jay Monkman as ARM architecture maintainer

2004-03-31	Ralf Corsepius <ralf_corsepius@rtems.org>

	* aclocal/rtems-top.m4: Remove ENDIF. 
	Remove with_target_subdir defaulting to '.'.
	* acinclude.m4: Add --includedir to sub_configure_args.

2004-02-17	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* Makefile.am: DIST_SUBDIRS = $(SUBDIRS).
	Rework dist-hook.
	* configure.ac: Reflect changes to RTEMS_ENABLE_RTEMSBSP.
	* aclocal/enable-rtemsbsp.m4: Add support for --disable-rtemsbsp.

2004-02-14	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* configure.ac: RTEMS_BUILD_CONFIG_SUBDIRS([testsuites/tools]).

2004-02-13	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* Makefile.am: Remove AUTOMAKE_FILES.
	Cleanup duplicate ACLOCAL_FILES.
	EXTRA_DIST += config-ml.in.
	DIST_SUBDIRS += make.
	Add dist-hook.
	* configure.ac: Add dist-bzip2 no-dist-gzip to AUTOMAKE_OPTIONS.

2004-02-12	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* bootstrap: RTEMS_CHECK_BSPDIR(RTEMS_BSP_FAMILY).

2004-02-12	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* VERSION: 4.6.99.0.
	* aclocal/version.m4: 4.6.99.0.

2004-02-06	Joel Sherrill <joel@OARcorp.com>

	* aclocal/bsp-alias.m4: Readd aliases for p4600 and p4650
	which were accidentally removed when purging the mips64orion.

2004-02-03	Ralf Corsepius <corsepiu@faw.uni-ulm.de>

	* aclocal/rpm.m4: Make osversions the default.
	* aclocal/rtems-top.m4: Remove check for version.m4 (Breaks make
	distcheck).

2004-02-01	Ralf Corsepius <corsepiu@faw.uni-ulm.de>


cpukit/libnetworking/libc/addr2ascii.c
/*
 * Copyright 1996 Massachusetts Institute of Technology
 *
 * Permission to use, copy, modify, and distribute this software and
 * its documentation for any purpose and without fee is hereby
 * granted, provided that both the above copyright notice and this
 * permission notice appear in all copies, that both the above
 * copyright notice and this permission notice appear in all
 * supporting documentation, and that the name of M.I.T. not be used
 * in advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.  M.I.T. makes
 * no representations about the suitability of this software for any
 * purpose.  It is provided "as is" without express or implied
 * warranty.
 * 
 * THIS SOFTWARE IS PROVIDED BY M.I.T. ``AS IS''.  M.I.T. DISCLAIMS
 * ALL EXPRESS OR IMPLIED WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. IN NO EVENT
 * SHALL M.I.T. BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	$ANA: addr2ascii.c,v 1.1 1996/06/13 18:41:46 wollman Exp $
 *
 *  $Id$
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif

#include <sys/types.h>
#include <sys/socket.h>

#include <errno.h>
#include <string.h>

#include <net/if_dl.h>
#include <netinet/in.h>
#include <arpa/inet.h>

/*-
 * Convert a network address from binary to printable numeric format.
 * This API is copied from INRIA's IPv6 implementation, but it is a
 * bit bogus in two ways:
 *
 *	1) There is no value in passing both an address family and
 *	   an address length; either one should imply the other,
 *	   or we should be passing sockaddrs instead.
 *	2) There should by contrast be /added/ a length for the buffer
 *	   that we pass in, so that programmers are spared the need to
 *	   manually calculate (read: ``guess'') the maximum length.
 *
 * Flash: the API is also the same in the NRL implementation, and seems to
 * be some sort of standard, so we appear to be stuck with both the bad
 * naming and the poor choice of arguments.
 */
char *
addr2ascii(af, addrp, len, buf)
	int af;
	const void *addrp;
	int len;		/* should be size_t XXX */
	char *buf;		/* XXX should pass length of buffer */
{
	static char staticbuf[64]; /* 64 for AF_LINK > 16 for AF_INET */

	if (!buf)
		buf = staticbuf;

	switch(af) {
	case AF_INET:
		if (len != sizeof(struct in_addr)) {
			errno = ENAMETOOLONG;
			return 0;
		}
		strcpy(buf, inet_ntoa(*(const struct in_addr *)addrp));
		break;

	case AF_LINK:
		if (len != sizeof(struct sockaddr_dl)) {
			errno = ENAMETOOLONG;
			return 0;
		}
		strcpy(buf, link_ntoa((const struct sockaddr_dl *)addrp));
		break;
			
	default:
		errno = EPROTONOSUPPORT;
		return 0;
	}
	return buf;
}
